### PR TITLE
#170947665 Create a two way trip request

### PR DESCRIPTION
**Pull request by T2Wil**
#### What does this PR do?
 This PR will help a user to request a trip from the manager.
#### Description of Task to be completed?
- Validate inputs
- Create requests and roles tables
- Write tests
- Write documentation

#### How should this be manually tested?
- run `git clone https://github.com/andela/knights-bn-backend.git` to clone the repo.
- run `cd knights-bn-backend` to go to the project directory.
- run `git checkout ft-two-way-trip-170947665` to switch to this branch.
- run `npm install` to install dependencies.
- create a Postgres database locally.
- run `sequelize db:migrate` to sync the newly created tables with your database
- run `sequelize db:seed:all` to initiate a default manager in the system
- create a .env file, paste in below variables and replace xxxx with the correct values: (Check code-knights `slack channel` to find the correct values) or add `MANAGER_PSWD=$2b$10$SO7hwwylEnHKZIv68w/ePOK4yybPnhXa5.tQSUBe.PSb6FBKyuVO.` to your already existing .env.

```
DB_USER='name of the user of the database'
DB_PASS='password of your Postgres'
DB_NAME='name of the database || Barefoot'
DB_TEST_NAME='name of the database for test || Barefoot_test'
NODE_ENV='development'
SECRETKEY='PrivateKey'
GOOGLE_CLIENT_ID = 'XXXXX'
GOOGLE_CLIENT_SECRET = 'XXXXX'
GOOGLE_CALLBACK_URL = 'XXXXX'
FCBK_CLIENT_ID = 'XXXXX'
FCBK_CLIENT_SECRET = 'XXXXX'
FCBK_CALLBACK_URL = 'XXXXX'
BN_API_KEY= 'XXXX'
MANAGER_PSWD= 'XXXX'
```
- run `npm run createTables` to create the database tables
- run `npm dev-start` to start the server.
- open postman and do sign up to acquire the token `POST` `http://localhost:4000/api/v1/auth/signup`
  and Fill in the required body request as below.
 ```
{
"firstName": "newiyonsenga",
"lastName": "Eric",
"gender": "male",
"passportNumber": "32825662",
"email": "ishitwill@gmail.com",
"password": "1#Niyonkuru"
}
```
- Do `POST` on `http://localhost:4000/api/v1/trips/returnTrip`

```     
{
	"origin": "Kigali",
	"destination": "Kampala",
	"departureDate": "2020-02-28",
	"returnDate": "2020-03-30",
	"accommodation": "XYZ campus",
	"reason": "Having fun",
	"passportNumber": "12345677"
}
```
[image](https://user-images.githubusercontent.com/30593186/74880953-0fd33380-5375-11ea-8eeb-ad0fd4a81220.png)

- You should see the output with a summary of the user's request and the request status.
```
{
    "message": "request created on success!",
    "origin": "Kigali",
    "destination": "Kampala",
    "departureDate": "2020-04-28",
    "returnDate": "2020-05-30",
    "reason": "Having fun",
    "requestId": 1,
    "requestType": "two way",
    "status": "pending"
}
```

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[link](https://www.pivotaltracker.com/story/show/170947665)

#### Screenshots
# 1. user signup
image

# 2. Populate request body and paste in the token in the header
image

# 3. Get a response body

image

### CODE REVIEW

**Review by Kagorora — APPROVED** [review]

**Comment by Moise1** [comment]
@T2Wil , Nice work!

**Review by Moise1 — APPROVED** [review]

**Review by Niyonsengaeric — APPROVED** [review]

**Review by EugeneMunya — APPROVED** [review]

File: src/db/models/request.js [hunk 1]
```diff
@@ -1,19 +1,17 @@
...
+    departureDate: DataTypes.DATEONLY,
+    returnDate: DataTypes.DATEONLY,
     reason: DataTypes.STRING,
-    destination: DataTypes.STRING,
     status: DataTypes.STRING,
-    departureDate: DataTypes.DATE,
-    returnDate: DataTypes.DATE
   }, {});
   Request.associate = (models) => {
-    Request.belongsTo(models.User, {
```
Comment (MuhanguziDavid) [inline src/db/models/request.js:13]: Why has this association been removed? It may be needed.

File: src/db/models/user.js [hunk 2]
```diff
@@ -12,10 +11,6 @@ module.exports = (sequelize, DataTypes) => {
     clientId: DataTypes.STRING
   }, {});
   User.associate = (models) => {
-    User.hasMany(models.Request, {
```
Comment (MuhanguziDavid) [inline src/db/models/user.js:52]: Why has this association been removed? It may be needed.

File: src/db/models/request.js [hunk 3]
```diff
@@ -1,19 +1,17 @@
 
 module.exports = (sequelize, DataTypes) => {
   const Request = sequelize.define('Request', {
-    userId: DataTypes.INTEGER,
+    managerId: DataTypes.INTEGER,
```
Comment (MuhanguziDavid) [inline src/db/models/request.js]: This should be a foreign key referring to a user who is a manager.

**Comment by MuhanguziDavid** [comment]
Nice work @T2Wil . When a new user is created, they should either be assigned as manager or requester. However, when I signup, I am not given a role, the `role` field is left as `NULL`. Please fix this.

File: src/db/models/request.js (hunk 1 above)
Comment (T2Wil) [reply src/db/models/request.js:13]: Hello @MuhanguziDavid ,
I have created a migration to update request tables and it is where I have created the required associations.

File: src/db/models/user.js (hunk 2 above)
Comment (T2Wil) [reply src/db/models/user.js:52]: I have added them in the new migrations I have created to update both the requests and users tables.

File: src/db/models/request.js (hunk 3 above)
Comment (T2Wil) [reply src/db/models/request.js]: Solved!

**Comment by T2Wil** [comment]
> Nice work @T2Wil . When a new user is created, they should either be assigned as manager or requester. However, when I signup, I am not given a role, the `role` field is left as `NULL`. Please fix this.

Thank you @MuhanguziDavid ,
I worked on the feedback and got it fixed!
# New user signing up
image

# Users' table
image

# Creating a request
image

# in request table
image

**Pull request merged**